### ISSUE TITLE
several instances with a single table may cause repeat consumption

### ISSUE THREAD

**Issue by li-zheng-hao** [dotnetcore/CAP#1271]
##  Environment

1. Storage: MongoDB
2. MQ: RabbitMQ
3. CAP:  6.2.0

## Problem

I have a service deployed with nearly 20 instances which storage messages in a single mongodb collection , consider this situation：

When a message is in the `Scheduled` or `Failed` state , the service need upgraded or restarted for some reasons, the failed message will be retried in all service instances when these instances startup at same time, which will lead to repeat consumption，the longer the message processing time, the greater the probability of repeat consumption.

Also, in normal situation, when a message fails for 3 times, it may be pulled and consumed by multiple instances when it is retried next time. 

I know CAP does not guarantee avoiding repeat consumption，But I still hope that the above problem can be solved

My current solution is to set up a separate table for each instance based on the hostname during initialization，but this brings up a new problem——the more instances I have, the more tables I have in my database, which makes me see a lot of CAP tables when managing the database (double of instances, publisher and receive table)

## The code and issue related to this problem

```c#
public async Task<IEnumerable<MediumMessage>> GetReceivedMessagesOfNeedRetry()
{
    var fourMinAgo = DateTime.Now.AddMinutes(-4);
    var collection = _database.GetCollection<ReceivedMessage>(_options.Value.ReceivedCollection);
    var queryResult = await collection
        .Find(x => x.Retries < _capOptions.Value.FailedRetryCount
                   && x.Added < fourMinAgo
                   && x.Version == _capOptions.Value.Version
                   && (x.StatusName == nameof(StatusName.Failed) ||
                       x.StatusName == nameof(StatusName.Scheduled)))
        .Limit(200)
        .ToListAsync().ConfigureAwait(false);
    return queryResult.Select(x => new MediumMessage
    {
        DbId = x.Id.ToString(),
        Origin = _serializer.Deserialize(x.Content)!,
        Retries = x.Retries,
        Added = x.Added
    }).ToList();
}
```

issue: https://github.com/dotnetcore/CAP/issues/170

**Comment by li-zheng-hao** [comment]
And why is it fixed for 4 minutes? Can this be a configuration parameter?

```c#
var fourMinAgo = DateTime.Now.AddMinutes(-4);
```

**Comment by yang-xiaodong** [comment]
Hi ,

Yes, this is a known issue, and it's the design in the past! Failure messages may be picked up by multiple nodes and executed repeatedly. So consumers should guarantee idempotence.

see : https://github.com/dotnetcore/CAP/pull/671#issuecomment-697071575

A feasible solution is to create a lock table. Before retrying, all instances acquire locks from the database, and the instances that acquire locks are retried. Currently, we are still evaluating the complexity and feasibility

**Comment by yang-xiaodong** [comment]
Fixed in version `7.1.0-preview-193202380` and relased to nuget a few minutes ago.
